aardvark-doc/src/document.rs

```diff
@@ -40,48 +64,185 @@ mod imp {
+                        glib::ControlFlow::Break
+                    }
+                ),
+            );
+        }
+
+        fn setup_loro_document(&self) {
+            let public_key = self.obj().service().public_key();
+            let obj = self.obj();
+            let doc = LoroDoc::new();
+            doc.set_peer_id({
+                // Take first 8 bytes of public key (32 bytes) to determine a unique "peer id"
+                // which is used to keep authors apart inside the text crdt.
+                //
+                // TODO(adz): This is strictly speaking not collision-resistant but we're limited
+                // here by the 8 bytes / 64 bit from the u64 `PeerId` type from Loro. In practice
+                // this should not really be a problem, but it would be nice if the Loro API would
+                // change some day.
+                let mut buf = [0u8; 8];
+                buf[..8].copy_from_slice(&public_key.as_bytes()[..8]);
+                u64::from_be_bytes(buf)
+            })
+            .expect("set peer id for new document");
+
+            let text = doc.get_text(TEXT_CONTAINER_ID);
+            doc.subscribe(
+                &text.id(),
+                Arc::new(clone!(
+                    #[weak]
+                    obj,
+                    move |loro_event| {
+                        let text_deltas = loro_event.events.into_iter().filter_map(|event| {
+                            if event.is_unknown {
+                                return None;
+                            }
+
+                            if let Diff::Text(loro_deltas) = event.diff {
+                                Some(loro_deltas)
+                            } else {
+                                None
+                            }
+                        });
+
+                        for commit in text_deltas {
+                            let mut index = 0;
+                            for delta in commit {
+                                match delta {
+                                    loro::TextDelta::Retain { retain, .. } => {
+                                        index += retain;
+                                    }
+                                    loro::TextDelta::Insert { insert, .. } => {
+                                        let len = insert.len();
+                                        obj.imp().emit_text_inserted(index as i32, insert);
+                                        index += len;
+                                    }
+                                    loro::TextDelta::Delete { delete } => {
+                                        obj.imp().emit_range_deleted(
+                                            index as i32,
+                                            (index + delete) as i32,
+                                        );
+                                    }
+                                }
+                            }
+                        }
```

Comment:
I like that this PR makes the code less "abstract" by reducing layers, but some doc-strings have been very useful as well in the previous code. Especially to explain the conversion around the delta types would be nice to mention here?